### DISCUSSION TITLE
[Feat] Annotation epochs should match accompanying alert state dates.

### DISCUSSION THREAD

**Discussion by calebtote** [grafana/grafana#36998] [Alerting]
### What Grafana version are you using?
v5.2.1

### What did you do?
Triggered an alert.

### What was the expected result?
Annotation created with the exact Epoch of the alert date.

### What happened instead?
With the recent update to convert epoch times to millisecond precision, it prompted me to make local changes that also converted the alert times to millisecond as well. This is required due to how we associate annotations with their accompanying alerts (i.e., we can search for which annotations were created from an alert by comparing the timestamps).

However, it seems the state changes, while close, aren't exact. I suspect we're doing a `time.Now().UnixNano() / int64(time.Millisecond)` or `timeNow().Round(time.Millisecond)` in a couple of places. When precision was only to the second, this was tolerable - however, with the millisecond precision, DB updates are 10-20ms apart from each other.

We should be pulling the epoch from the annotation (or the state date from the alert, whichever comes first) to ensure the times are exact.

While this only affects my custom tooling that correlates alerts with their accompanying annotations, I can foresee this being a larger issue down the road as alerts and annotations become more natively intermingled.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
